### PR TITLE
Make Host.StartAsync cancellable by app lifetime

### PR DESCRIPTION
**Pull request by poke**
This fixes aspnet/AspNetCore#10137 (which still has to be moved to this repo).

Originally, `Host.StartAsync` would only start the web host which started very quickly. With the generic host and custom hosted services, it becomes more likely that the host startup process will take longer,
especially when hosted services need to do more work during startup (e.g. migrations). This increases the need for the startup process to be cancellable.

Hosts already have a concept of being externally cancellable, e.g. when running inside of the console or as a Windows service. This is done through the host application lifetime which can gracefully shut down the
application.

This change integrates the host application lifetime into the startup process to make sure that a stopped application lifetime will also cancel the startup process. That way, the application can shut down
immediately instead of having to wait for the startup to complete in order to terminate.

/cc @anurse, @Tratcher

### CODE REVIEW

File: src/Hosting/Hosting/test/Internal/HostTests.cs [hunk 1]
```diff
@@ -731,9 +818,6 @@ public async Task HostedServiceCanInjectApplicationLifetime()
                    })
                    .Build())
             {
-                var lifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();
-                lifetime.StopApplication();
```
Comment (poke) [inline src/Hosting/Hosting/test/Internal/HostTests.cs]: I’m not sure why this was there in this test. According to the name, this just checks that the lifetime can be injected into services, which will not require the application to be stopped.

File: src/Hosting/Hosting/test/Internal/HostTests.cs [hunk 2]
```diff
@@ -709,13 +796,13 @@ public async Task HostStopApplicationDoesNotFireStopOnHostedService()
...
-
                 await host.StartAsync();
 
                 Assert.Equal(0, stoppingCalls);
                 Assert.Equal(0, disposingCalls);
+
+                var lifetime = host.Services.GetRequiredService<IHostApplicationLifetime>();
+                lifetime.StopApplication();
```
Comment (poke) [inline src/Hosting/Hosting/test/Internal/HostTests.cs]: Same as below, not sure what the original purpose of this was, but I left it in here to make sure that the `StopApplication` will not call the `StopAsync` of the hosted service.

File: src/Hosting/Hosting/src/Internal/Host.cs [hunk 3]
```diff
@@ -39,19 +39,24 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
         {
             _logger.Starting();
 
-            await _hostLifetime.WaitForStartAsync(cancellationToken);
+            using (var combinedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _applicationLifetime.ApplicationStopping))
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/Internal/Host.cs]: Can we use C#8 in Extensions?
```suggestion
            using var combinedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _applicationLifetime.ApplicationStopping);
```
Less `{ }`

File: src/Hosting/Hosting/test/Internal/HostTests.cs [hunk 4]
```diff
@@ -266,6 +266,93 @@ public async Task StartCanBeCancelled()
...
+            var hostedService = new AsyncFakeHostedService();
+            var builder = CreateBuilder().ConfigureServices(services =>
+            {
+                services.AddSingleton<IHostedService>(hostedService);
+            });
+
+            using (var host = builder.Build())
+            {
+                var applicationLifetime = host.Services.GetService<IHostApplicationLifetime>();
+                var startTask = Task.Run(() => host.StartAsync());
```
Comment (Tratcher) [inline src/Hosting/Hosting/test/Internal/HostTests.cs]: Task.Run isn't needed here, AsyncFakeHostedService.StartAsync doesn't block sync.

File: src/Hosting/Hosting/test/Internal/HostTests.cs (hunk 2 above)
Comment (Tratcher) [reply src/Hosting/Hosting/test/Internal/HostTests.cs]: Move above the Asserts?

File: src/Hosting/Hosting/src/Internal/Host.cs (hunk 3 above)
Comment (poke) [reply src/Hosting/Hosting/src/Internal/Host.cs]: I had that in a draft and then opted for a normal using block… 😅 

I don’t have a preference for either though, so I can change that back.

File: src/Hosting/Hosting/test/Internal/HostTests.cs (hunk 4 above)
Comment (poke) [reply src/Hosting/Hosting/test/Internal/HostTests.cs]: Oh right, that’s a left-over from an early version where I was using that sync `ManualResetEvent` that the other tests use.

**Comment by Tratcher** [comment]
Thanks

**Comment by PureKrome** [comment]
Thanks @poke for the PR and the @Tratcher and @anurse for a quick turn around. Awesome effort :)

Just to confirm @Tratcher - this will be available in `Preview6` to test/check?

**Comment by Tratcher** [comment]
@PureKrome correct. It should be available in the Master preview6 nightly SDK some time tomorrow. https://github.com/dotnet/core-sdk

**Pull request merged**